File: sdks/py_sdk/sw4rm/activity_buffer.py
from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Dict, Optional, Any, List
from collections import deque

from . import constants as C
from .exceptions import BufferFullError
from .persistence import (
    ActivityBufferLoadError,
    PersistenceBackend,
    JSONFilePersistence,
    PersistentActivityRecord,
)


def is_terminal_state(state: int) -> bool:
    """Check if an envelope state is terminal (no further transitions expected).

    Terminal states are: FULFILLED, REJECTED, FAILED, TIMED_OUT

    Args:
        state: Envelope state value

    Returns:
        True if state is terminal, False otherwise
    """
    return state in (
        C.FULFILLED_ENVELOPE,
        C.REJECTED_ENVELOPE,
        C.FAILED_ENVELOPE,
        C.TIMED_OUT_ENVELOPE,
    )


@dataclass
class ActivityRecord:
    message_id: str
    direction: str  # "in" | "out"
    envelope: Dict[str, Any]
    ts_ms: int = field(default_factory=lambda: int(time.time() * 1000))
    ack_stage: int = C.ACK_STAGE_UNSPECIFIED
    error_code: int = C.ERROR_CODE_UNSPECIFIED
    ack_note: str = ""

    def ack(self, stage: int, error_code: int = C.ERROR_CODE_UNSPECIFIED, note: str = "") -> None:
        self.ack_stage = stage
        self.error_code = error_code
        self.ack_note = note

    @property
    def state(self) -> int:
        """Get envelope state from the envelope dict."""
        return self.envelope.get("state", C.ENVELOPE_STATE_UNSPECIFIED)

    @property
    def idempotency_token(self) -> str:
        """Get idempotency token from the envelope dict."""
        return self.envelope.get("idempotency_token", "")

    @property
    def correlation_id(self) -> str:
        """Get correlation ID from the envelope dict."""
        return self.envelope.get("correlation_id", "")


class ActivityBuffer:
    """In-memory activity buffer.

    Capacity is enforced per spec 10.1: registrations beyond max_items are
    rejected with BufferFullError rather than silently evicted (no pluggable
    eviction API; fp-ds-certification#F6 remove branch).

    Tracks inbound/outbound envelopes by message_id and records ACK progression.
    Supports Three-ID model: deduplication via idempotency_token, workflow linking
    via correlation_id, and per-attempt tracking via message_id.

    Not thread-safe; callers should synchronize if used across threads.
    """

    def __init__(
        self,
        *,
        max_items: int = 10000,
        dedup_window_s: int = 3600
    ) -> None:
        self._by_id: Dict[str, ActivityRecord] = {}
        self._by_idempotency_token: Dict[str, str] = {}  # token -> message_id
        self._order: deque[str] = deque()
        self._max_items = max_items
        self.dedup_window_s = dedup_window_s  # Default: 1 hour per spec §11.1

    def _check_capacity(self) -> None:
        """Reject new entries when buffer is full per spec §10.1.

        Raises:
            BufferFullError: When buffer is at capacity
        """
        if len(self._by_id) >= self._max_items:
            raise BufferFullError(
                f"Activity buffer is full (max {self._max_items} items)",
                current_size=len(self._by_id),
                max_size=self._max_items,
            )

    def _cleanup_expired_dedup_entries(self) -> None:
        """Remove idempotency tokens for records outside dedup window."""
        now_ms = int(time.time() * 1000)
        window_ms = self.dedup_window_s * 1000

        # Find expired tokens
        expired_tokens = []
        for token, message_id in self._by_idempotency_token.items():
            rec = self._by_id.get(message_id)
            if rec and (now_ms - rec.ts_ms) > window_ms:
                expired_tokens.append(token)

        # Remove expired mappings
        for token in expired_tokens:
            del self._by_idempotency_token[token]

    def record_incoming(self, envelope: Dict[str, Any]) -> ActivityRecord:
        mid = str(envelope.get("message_id"))
        if mid not in self._by_id:
            self._check_capacity()
            self._order.append(mid)
        rec = ActivityRecord(message_id=mid, direction="in", envelope=envelope)
        self._by_id[mid] = rec

        # Track idempotency token for deduplication
        token = rec.idempotency_token
        if token:
            self._by_idempotency_token[token] = mid

        self._cleanup_expired_dedup_entries()
        return rec

    def record_outgoing(self, envelope: Dict[str, Any]) -> ActivityRecord:
        mid = str(envelope.get("message_id"))
        if mid not in self._by_id:
            self._check_capacity()
            self._order.append(mid)
        rec = ActivityRecord(message_id=mid, direction="out", envelope=envelope)
        self._by_id[mid] = rec

        # Track idempotency token for deduplication
        token = rec.idempotency_token
        if token:
            self._by_idempotency_token[token] = mid

        self._cleanup_expired_dedup_entries()
        return rec

    def ack(self, ack: Dict[str, Any]) -> Optional[ActivityRecord]:
        target = str(ack.get("ack_for_message_id"))
        rec = self._by_id.get(target)
        if rec:
            rec.ack(
                stage=int(ack.get("ack_stage", C.ACK_STAGE_UNSPECIFIED)),
                error_code=int(ack.get("error_code", C.ERROR_CODE_UNSPECIFIED)),
                note=str(ack.get("note", "")),
            )
        return rec

    def get(self, message_id: str) -> Optional[ActivityRecord]:
        return self._by_id.get(message_id)

    def unacked(self) -> List[ActivityRecord]:
        return [r for r in self._by_id.values() if r.ack_stage in (C.ACK_STAGE_UNSPECIFIED, C.RECEIVED, C.READ)]

    def recent(self, n: int = 50) -> List[ActivityRecord]:
        ids = list(self._order)[-n:] if n > 0 else []
        return [self._by_id[i] for i in ids if i in self._by_id]

    def update_state(self, message_id: str, new_state: int) -> Optional[ActivityRecord]:
        """Update the envelope state for a specific message.

        Args:
            message_id: The message_id of the envelope to update
            new_state: New envelope state (see constants.EnvelopeState)

        Returns:
            Updated ActivityRecord if found, None otherwise
        """
        rec = self._by_id.get(message_id)
        if rec:
            rec.envelope["state"] = new_state
        return rec

    def get_by_idempotency_token(self, token: str) -> Optional[ActivityRecord]:
        """Get an activity record by its idempotency token.

        Used for deduplication: check if an operation with this token
        has already been processed within the deduplication window.

        Args:
            token: Idempotency token to look up

        Returns:
            ActivityRecord if found and within dedup window, None otherwise

        Example:
            >>> # Check for duplicate before processing
            >>> existing = buffer.get_by_idempotency_token(token)
            >>> if existing and not is_terminal_state(existing.state):
            ...     # Duplicate in progress, reject or return cached result
            ...     return existing
        """
        self._cleanup_expired_dedup_entries()
        message_id = self._by_idempotency_token.get(token)
        if message_id:
            return self._by_id.get(message_id)
        return None


class PersistentActivityBuffer:
    """Activity buffer with persistent storage across restarts.

    Supports multiple persistence backends (JSON file, SQLite, etc.)
    and provides reconciliation on startup to restore previous state.
    Includes Three-ID support for deduplication and workflow tracking.
    """

    def __init__(
        self,
        *,
        max_items: int = 10000,
        persistence: Optional[PersistenceBackend] = None,
        dedup_window_s: int = 3600,
        load_failure_mode: str = "raise",
    ):
        self._by_id: Dict[str, PersistentActivityRecord] = {}
        self._by_idempotency_token: Dict[str, str] = {}  # token -> message_id
        self._order: deque[str] = deque()
        self._max_items = max_items
        self._persistence = persistence or JSONFilePersistence()
        self.dedup_window_s = dedup_window_s
        # "raise" (default): a corrupted persistence file fails startup loudly.
        # "empty": catch and start empty, logging a degraded-mode warning —
        # opt-in escape hatch for best-effort consumers; never silent.
        if load_failure_mode not in ("raise", "empty"):
            raise ValueError(
                f"load_failure_mode must be 'raise' or 'empty', got {load_failure_mode!r}"
            )
        self.load_failure_mode = load_failure_mode
        self._dirty = False  # Track if we need to save

        # Load existing data on initialization
        self._load_from_persistence()

    def _load_from_persistence(self) -> None:
        """Load activity records from persistent storage.

        A missing persistence file is a fresh start (no error). An
        unreadable/corrupted file raises ActivityBufferLoadError by
        default; with load_failure_mode="empty" it logs loudly and starts
        degraded (never silently).
        """
        try:
            records_data, order = self._persistence.load_records()
        except ActivityBufferLoadError:
            if self.load_failure_mode == "empty":
                import logging

                logging.getLogger(__name__).exception(
                    "[ActivityBuffer] CORRUPTED persistence; starting DEGRADED "
                    "with an empty buffer — recovery state was not restored"
                )
                self._by_id = {}
                self._by_idempotency_token = {}
                self._order = deque()
                return
            raise

        # Reconstruct activity records
        self._by_id = {}
        self._by_idempotency_token = {}
        for message_id, data in records_data.items():
            rec = PersistentActivityRecord.from_dict(data)
            self._by_id[message_id] = rec
            # Rebuild idempotency token index
            token = rec.envelope.get("idempotency_token", "")
            if token:
                self._by_idempotency_token[token] = message_id

        self._order = deque(order)
        if len(self._by_id) > self._max_items:
            self._check_capacity()

        print(f"[ActivityBuffer] Loaded {len(self._by_id)} records from persistence")

    def _save_to_persistence(self) -> None:
        """Save current state to persistent storage.

        Durability contract: save failures raise — silently swallowed save
        errors made the crash-recovery claim false.
        """
        if not self._dirty:
            return

        records_data = {mid: rec.to_dict() for mid, rec in self._by_id.items()}
        self._persistence.save_records(records_data, list(self._order))
        self._dirty = False

    def _check_capacity(self) -> None:
        """Reject new entries when buffer is full per spec §10.1.

        Raises:
            BufferFullError: When buffer is at capacity
        """
        if len(self._by_id) >= self._max_items:
            raise BufferFullError(
                f"Activity buffer is full (max {self._max_items} items)",
                current_size=len(self._by_id),
                max_size=self._max_items,
            )

    def _cleanup_expired_dedup_entries(self) -> None:
        """Remove idempotency tokens for records outside dedup window."""
        now_ms = int(time.time() * 1000)
        window_ms = self.dedup_window_s * 1000

        # Find expired tokens
        expired_tokens = []
        for token, message_id in self._by_idempotency_token.items():
            rec = self._by_id.get(message_id)
            if rec and (now_ms - rec.ts_ms) > window_ms:
                expired_tokens.append(token)

        # Remove expired mappings
        for token in expired_tokens:
            del self._by_idempotency_token[token]

        if expired_tokens:
            self._dirty = True

    def record_incoming(self, envelope: Dict[str, Any]) -> PersistentActivityRecord:
        """Record an incoming message envelope."""
        mid = str(envelope.get("message_id"))
        if mid not in self._by_id:
            self._check_capacity()
            self._order.append(mid)
        rec = PersistentActivityRecord(message_id=mid, direction="in", envelope=envelope)

        self._by_id[mid] = rec

        # Track idempotency token for deduplication
        token = rec.envelope.get("idempotency_token", "")
        if token:
            self._by_idempotency_token[token] = mid

        self._cleanup_expired_dedup_entries()
        self._dirty = True

        return rec

    def record_outgoing(self, envelope: Dict[str, Any]) -> PersistentActivityRecord:
        """Record an outgoing message envelope."""
        mid = str(envelope.get("message_id"))
        if mid not in self._by_id:
            self._check_capacity()
            self._order.append(mid)
        rec = PersistentActivityRecord(message_id=mid, direction="out", envelope=envelope)

        self._by_id[mid] = rec

        # Track idempotency token for deduplication
        token = rec.envelope.get("idempotency_token", "")
        if token:
            self._by_idempotency_token[token] = mid

        self._cleanup_expired_dedup_entries()
        self._dirty = True

        return rec

    def ack(self, ack: Dict[str, Any]) -> Optional[PersistentActivityRecord]:
        """Process an acknowledgment for a previously recorded message."""
        target = str(ack.get("ack_for_message_id"))
        rec = self._by_id.get(target)
        
        if rec:
            rec.ack(
                stage=int(ack.get("ack_stage", C.ACK_STAGE_UNSPECIFIED)),
                error_code=int(ack.get("error_code", C.ERROR_CODE_UNSPECIFIED)),
                note=str(ack.get("note", "")),
            )
            self._dirty = True
            
        return rec

    def get(self, message_id: str) -> Optional[PersistentActivityRecord]:
        """Get an activity record by message ID."""
        return self._by_id.get(message_id)

    def unacked(self) -> List[PersistentActivityRecord]:
        """Get all records that haven't been fully acknowledged."""
        return [r for r in self._by_id.values() 
                if r.ack_stage in (C.ACK_STAGE_UNSPECIFIED, C.RECEIVED, C.READ)]

    def recent(self, n: int = 50) -> List[PersistentActivityRecord]:
        """Get the N most recent activity records."""
        ids = list(self._order)[-n:] if n > 0 else []
        return [self._by_id[i] for i in ids if i in self._by_id]

    def flush(self) -> None:
        """Force save to persistent storage."""
        self._save_to_persistence()

    def reconcile(self) -> List[PersistentActivityRecord]:
        """Return unacked outgoing messages that may need retry/reconciliation."""
        unacked_outgoing = [
            rec for rec in self.unacked()
            if rec.direction == "out"
        ]
        return unacked_outgoing

    def update_state(self, message_id: str, new_state: int) -> Optional[PersistentActivityRecord]:
        """Update the envelope state for a specific message.

        Args:
            message_id: The message_id of the envelope to update
            new_state: New envelope state (see constants.EnvelopeState)

        Returns:
            Updated PersistentActivityRecord if found, None otherwise
        """
        rec = self._by_id.get(message_id)
        if rec:
            rec.envelope["state"] = new_state
            self._dirty = True
        return rec

    def get_by_idempotency_token(self, token: str) -> Optional[PersistentActivityRecord]:
        """Get an activity record by its idempotency token.

        Used for deduplication: check if an operation with this token
        has already been processed within the deduplication window.

        Args:
            token: Idempotency token to look up

        Returns:
            PersistentActivityRecord if found and within dedup window, None otherwise
        """
        self._cleanup_expired_dedup_entries()
        message_id = self._by_idempotency_token.get(token)
        if message_id:
            return self._by_id.get(message_id)
        return None

    def clear(self) -> None:
        """Clear all activity records and persistent storage."""
        self._by_id.clear()
        self._by_idempotency_token.clear()
        self._order.clear()
        self._persistence.clear()
        self._dirty = False

    def __enter__(self):
        """Context manager entry."""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit - ensure data is saved."""
        self._save_to_persistence()

File: sdks/rust_sdk/src/activity_buffer.rs
use crate::constants::*;
use crate::persistence::{PersistenceBackend, JsonFilePersistence, PersistentActivityRecord};
use crate::{Error, Result};
use std::collections::HashMap;
use std::sync::{Arc, RwLock};
use chrono::Utc;

/// Activity record for tracking message flow
#[derive(Debug, Clone)]
pub struct ActivityRecord {
    pub message_id: String,
    pub direction: String, // "in" | "out"
    pub envelope: serde_json::Value,
    pub ts_ms: i64,
    pub ack_stage: i32,
    pub error_code: i32,
    pub ack_note: String,
}

impl ActivityRecord {
    pub fn new(message_id: String, direction: String, envelope: serde_json::Value) -> Self {
        Self {
            message_id,
            direction,
            envelope,
            ts_ms: Utc::now().timestamp_millis(),
            ack_stage: ack_stage::ACK_STAGE_UNSPECIFIED,
            error_code: error_code::ERROR_CODE_UNSPECIFIED,
            ack_note: String::new(),
        }
    }

    pub fn ack(&mut self, stage: i32, error_code: i32, note: String) {
        self.ack_stage = stage;
        self.error_code = error_code;
        self.ack_note = note;
    }
}

/// In-memory activity buffer with simple reconciliation
#[derive(Debug)]
pub struct ActivityBuffer {
    by_id: HashMap<String, ActivityRecord>,
    order: Vec<String>,
    max_items: usize,
}

impl ActivityBuffer {
    pub fn new(max_items: usize) -> Self {
        Self {
            by_id: HashMap::new(),
            order: Vec::new(),
            max_items,
        }
    }

    fn prune_if_needed(&mut self) {
        while self.order.len() > self.max_items {
            let oldest = self.order.remove(0);
            self.by_id.remove(&oldest);
        }
    }

    pub fn record_incoming(&mut self, envelope: serde_json::Value) -> Result<ActivityRecord> {
        let message_id = envelope
            .get("message_id")
            .and_then(|v| v.as_str())
            .ok_or_else(|| Error::InvalidEnvelope("Missing message_id".to_string()))?
            .to_string();

        let record = ActivityRecord::new(message_id.clone(), "in".to_string(), envelope);
        self.by_id.insert(message_id.clone(), record.clone());
        self.order.push(message_id);
        self.prune_if_needed();
        
        Ok(record)
    }

    pub fn record_outgoing(&mut self, envelope: serde_json::Value) -> Result<ActivityRecord> {
        let message_id = envelope
            .get("message_id")
            .and_then(|v| v.as_str())
            .ok_or_else(|| Error::InvalidEnvelope("Missing message_id".to_string()))?
            .to_string();

        let record = ActivityRecord::new(message_id.clone(), "out".to_string(), envelope);
        self.by_id.insert(message_id.clone(), record.clone());
        self.order.push(message_id);
        self.prune_if_needed();
        
        Ok(record)
    }

    pub fn ack(&mut self, ack: &serde_json::Value) -> Option<&mut ActivityRecord> {
        let target = ack
            .get("ack_for_message_id")
            .and_then(|v| v.as_str())?
            .to_string();

        let record = self.by_id.get_mut(&target)?;
        
        let stage = ack
            .get("ack_stage")
            .and_then(|v| v.as_i64())
            .unwrap_or(ack_stage::ACK_STAGE_UNSPECIFIED as i64) as i32;
        
        let error_code = ack
            .get("error_code")
            .and_then(|v| v.as_i64())
            .unwrap_or(error_code::ERROR_CODE_UNSPECIFIED as i64) as i32;
        
        let note = ack
            .get("note")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();

        record.ack(stage, error_code, note);
        Some(record)
    }

    pub fn get(&self, message_id: &str) -> Option<&ActivityRecord> {
        self.by_id.get(message_id)
    }

    pub fn unacked(&self) -> Vec<&ActivityRecord> {
        self.by_id
            .values()
            .filter(|r| matches!(r.ack_stage, 
                ack_stage::ACK_STAGE_UNSPECIFIED | 
                ack_stage::RECEIVED | 
                ack_stage::READ
            ))
            .collect()
    }

    pub fn recent(&self, n: usize) -> Vec<&ActivityRecord> {
        let start = if self.order.len() > n { self.order.len() - n } else { 0 };
        
        self.order[start..]
            .iter()
            .filter_map(|id| self.by_id.get(id))
            .collect()
    }
}

/// Thread-safe persistent activity buffer
pub struct PersistentActivityBuffer {
    inner: Arc<RwLock<PersistentActivityBufferInner>>,
}

struct PersistentActivityBufferInner {
    by_id: HashMap<String, PersistentActivityRecord>,
    order: Vec<String>,
    max_items: usize,
    persistence: Box<dyn PersistenceBackend>,
    dirty: bool,
}

impl PersistentActivityBuffer {
    pub fn new(max_items: usize, persistence: Option<Box<dyn PersistenceBackend>>) -> Result<Self> {
        let persistence = persistence.unwrap_or_else(|| Box::new(JsonFilePersistence::default()));
        
        let mut inner = PersistentActivityBufferInner {
            by_id: HashMap::new(),
            order: Vec::new(),
            max_items,
            persistence,
            dirty: false,
        };

        // Load existing data
        inner.load_from_persistence()?;

        Ok(Self {
            inner: Arc::new(RwLock::new(inner)),
        })
    }

    pub fn record_incoming(&self, envelope: serde_json::Value) -> Result<PersistentActivityRecord> {
        let mut inner = self.inner.write().map_err(|_| Error::Internal("Lock poisoned".to_string()))?;
        
        let message_id = envelope
            .get("message_id")
            .and_then(|v| v.as_str())
            .ok_or_else(|| Error::InvalidEnvelope("Missing message_id".to_string()))?
            .to_string();

        let record = PersistentActivityRecord::new(message_id.clone(), "in".to_string(), envelope);
        inner.by_id.insert(message_id.clone(), record.clone());
        inner.order.push(message_id);
        inner.prune_if_needed();
        inner.dirty = true;
        
        Ok(record)
    }

    pub fn record_outgoing(&self, envelope: serde_json::Value) -> Result<PersistentActivityRecord> {
        let mut inner = self.inner.write().map_err(|_| Error::Internal("Lock poisoned".to_string()))?;
        
        let message_id = envelope
            .get("message_id")
            .and_then(|v| v.as_str())
            .ok_or_else(|| Error::InvalidEnvelope("Missing message_id".to_string()))?
            .to_string();

        let record = PersistentActivityRecord::new(message_id.clone(), "out".to_string(), envelope);
        inner.by_id.insert(message_id.clone(), record.clone());
        inner.order.push(message_id);
        inner.prune_if_needed();
        inner.dirty = true;
        
        Ok(record)
    }

    pub fn ack(&self, ack: &serde_json::Value) -> Result<Option<PersistentActivityRecord>> {
        let mut inner = self.inner.write().map_err(|_| Error::Internal("Lock poisoned".to_string()))?;
        
        let target = ack
            .get("ack_for_message_id")
            .and_then(|v| v.as_str())
            .ok_or_else(|| Error::InvalidEnvelope("Missing ack_for_message_id".to_string()))?
            .to_string();

        if let Some(record) = inner.by_id.get_mut(&target) {
            let stage = ack
                .get("ack_stage")
                .and_then(|v| v.as_i64())
                .unwrap_or(ack_stage::ACK_STAGE_UNSPECIFIED as i64) as i32;
            
            let error_code = ack
                .get("error_code")
                .and_then(|v| v.as_i64())
                .unwrap_or(error_code::ERROR_CODE_UNSPECIFIED as i64) as i32;
            
            let note = ack
                .get("note")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string();

            record.ack(stage, error_code, note);
            let result = record.clone();
            
            // Release the mutable borrow before accessing inner.dirty
            let _ = record;
            inner.dirty = true;
            
            Ok(Some(result))
        } else {
            Ok(None)
        }
    }

    pub fn get(&self, message_id: &str) -> Result<Option<PersistentActivityRecord>> {
        let inner = self.inner.read().map_err(|_| Error::Internal("Lock poisoned".to_string()))?;
        Ok(inner.by_id.get(message_id).cloned())
    }

    pub fn unacked(&self) -> Result<Vec<PersistentActivityRecord>> {
        let inner = self.inner.read().map_err(|_| Error::Internal("Lock poisoned".to_string()))?;
        let records = inner
            .by_id
            .values()
            .filter(|r| matches!(r.ack_stage, 
                ack_stage::ACK_STAGE_UNSPECIFIED | 
                ack_stage::RECEIVED | 
                ack_stage::READ
            ))
            .cloned()
            .collect();
        Ok(records)
    }

    pub fn recent(&self, n: usize) -> Result<Vec<PersistentActivityRecord>> {
        let inner = self.inner.read().map_err(|_| Error::Internal("Lock poisoned".to_string()))?;
        let start = if inner.order.len() > n { inner.order.len() - n } else { 0 };
        
        let records = inner.order[start..]
            .iter()
            .filter_map(|id| inner.by_id.get(id))
            .cloned()
            .collect();
        
        Ok(records)
    }

    pub fn flush(&self) -> Result<()> {
        let mut inner = self.inner.write().map_err(|_| Error::Internal("Lock poisoned".to_string()))?;
        inner.save_to_persistence()
    }

    pub fn reconcile(&self) -> Result<Vec<PersistentActivityRecord>> {
        let unacked = self.unacked()?;
        let outgoing_unacked = unacked
            .into_iter()
            .filter(|r| r.direction == "out")
            .collect();
        
        Ok(outgoing_unacked)
    }

    pub fn clear(&self) -> Result<()> {
        let mut inner = self.inner.write().map_err(|_| Error::Internal("Lock poisoned".to_string()))?;
        inner.by_id.clear();
        inner.order.clear();
        inner.persistence.clear()?;
        inner.dirty = false;
        Ok(())
    }
}

impl PersistentActivityBufferInner {
    fn load_from_persistence(&mut self) -> Result<()> {
        match self.persistence.load_records() {
            Ok((records_data, order)) => {
                self.by_id = records_data;
                self.order = order;
                self.prune_if_needed();
                tracing::info!("Loaded {} records from persistence", self.by_id.len());
                Ok(())
            }
            Err(e) => {
                tracing::warn!("Failed to load from persistence: {}", e);
                self.by_id.clear();
                self.order.clear();
                Ok(())
            }
        }
    }

    fn save_to_persistence(&mut self) -> Result<()> {
        if !self.dirty {
            return Ok(());
        }

        self.persistence.save_records(self.by_id.clone(), self.order.clone())?;
        self.dirty = false;
        Ok(())
    }

    fn prune_if_needed(&mut self) {
        while self.order.len() > self.max_items {
            let oldest = self.order.remove(0);
            self.by_id.remove(&oldest);
            self.dirty = true;
        }
    }
}

impl Drop for PersistentActivityBuffer {
    fn drop(&mut self) {
        // Best effort save on drop
        if let Ok(mut inner) = self.inner.write() {
            let _ = inner.save_to_persistence();
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::persistence::JsonFilePersistence;
    use serde_json::json;
    use tempfile::NamedTempFile;
    use crate::constants;

    #[test]
    fn test_activity_buffer_creation() {
        let buffer = ActivityBuffer::new(100);
        assert_eq!(buffer.capacity(), 100);
        assert_eq!(buffer.len(), 0);
        assert!(buffer.is_empty());
        assert!(buffer.unacked().unwrap().is_empty());
    }

    #[test]
    fn test_activity_buffer_record_incoming() {
        let mut buffer = ActivityBuffer::new(10);
        
        let message = json!({
            "message_id": "msg-1",
            "producer_id": "test-agent",
            "message_type": constants::message_type::DATA,
            "payload": "test data"
        });

        let record = buffer.record_incoming(message.clone()).unwrap();
        
        assert_eq!(record.message_id, "msg-1");
        assert_eq!(record.producer_id, "test-agent");
        assert_eq!(record.direction, "in");
        assert_eq!(record.ack_stage, constants::ack_stage::ACK_STAGE_UNSPECIFIED);
        assert_eq!(buffer.len(), 1);
        assert!(!buffer.is_empty());
    }

    #[test]
    fn test_activity_buffer_ack_processing() {
        let mut buffer = ActivityBuffer::new(10);
        
        // Record incoming message
        let message = json!({
            "message_id": "msg-ack-test",
            "producer_id": "test-agent",
            "message_type": constants::message_type::DATA,
            "payload": "ack test data"
        });

        buffer.record_incoming(message).unwrap();
        
        // Send ACK
        let ack = json!({
            "ack_for_message_id": "msg-ack-test",
            "ack_stage": constants::ack_stage::RECEIVED,
            "error_code": constants::error_code::ERROR_CODE_UNSPECIFIED,
            "note": "Message received successfully"
        });

        let ack_result = buffer.ack(&ack).unwrap();
        assert!(ack_result.is_some());
        
        let updated_record = ack_result.unwrap();
        assert_eq!(updated_record.ack_stage, constants::ack_stage::RECEIVED);
        assert_eq!(updated_record.ack_note, "Message received successfully");
    }

    #[test]
    fn test_persistent_activity_buffer_creation() {
        let temp_file = NamedTempFile::new().unwrap();
        let persistence = Box::new(JsonFilePersistence::new(temp_file.path()));
        
        let buffer = PersistentActivityBuffer::new(50, Some(persistence));
        assert!(buffer.is_ok());
        
        let buffer = buffer.unwrap();
        assert_eq!(buffer.capacity(), 50);
        assert!(buffer.is_empty());
    }

    #[test]
    fn test_activity_buffer_capacity_management() {
        let mut buffer = ActivityBuffer::new(3);
        
        // Add more messages than capacity
        for i in 1..=5 {
            let message = json!({
                "message_id": format!("capacity-{}", i),
                "producer_id": "test-agent",
                "message_type": constants::message_type::DATA,
                "payload": format!("data-{}", i)
            });
            buffer.record_incoming(message).unwrap();
        }

        // Should only keep the most recent 3 messages
        assert_eq!(buffer.len(), 3);
        
        let all_records = buffer.recent(10).unwrap();
        assert_eq!(all_records.len(), 3);
        assert_eq!(all_records[0].message_id, "capacity-5");
        assert_eq!(all_records[2].message_id, "capacity-3");
    }

    #[test]
    fn test_activity_buffer_error_handling() {
        let mut buffer = ActivityBuffer::new(10);
        
        // Test with invalid message (missing required fields)
        let invalid_message = json!({"incomplete": "message"});
        let result = buffer.record_incoming(invalid_message);
        assert!(result.is_err());

        // Test ACK for non-existent message
        let ack_for_missing = json!({
            "ack_for_message_id": "does-not-exist",
            "ack_stage": constants::ack_stage::RECEIVED,
            "error_code": constants::error_code::ERROR_CODE_UNSPECIFIED,
            "note": "test"
        });
        
        let ack_result = buffer.ack(&ack_for_missing);
        assert!(ack_result.is_none());
    }
}